### PR TITLE
Do not run UDFs when the partition is empty.

### PR DESCRIPTION
**Pull request by firestarman**
When the partition is empty, do not launch the Python runners and Python processes which are expensive and useless, and return the input iterator directly instead.

Filed this PR to fix the udf related failures in IT tests due to IPC error. This is not a fix for the root cause, but it still makes sense. 

It looks like the cuDF 'ArrowIPCTableWriter' does not handle the empty partiton case correctly. So the Python side reports the IPC error. 

The IPC error should have been there for a while, however we just ignored it due to not read the control data before. But now is exposed by the PR https://github.com/NVIDIA/spark-rapids/pull/1235. Anyway will continue investigating it.

Also reduce the data size for cudf tests to save running time.

Signed-off-by: Firestarman <firestarmanllc@gmail.com>

### CODE REVIEW

**Comment by firestarman** [comment]
build

File: integration_tests/src/main/python/udf_cudf_test.py [hunk 1]
```diff
@@ -45,7 +45,7 @@
 
 small_data = [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)]
 
-large_data = list(map(lambda i: (i, i/1.0), range(1, 5000))) * 2
+large_data = list(map(lambda i: (i, i/1.0), range(1, 512))) * 2
```
Comment (revans2) [inline integration_tests/src/main/python/udf_cudf_test.py:48]: This change appears to be unrelated to running UDFs with an empty partition.  Is there some reason I am missing?

**Review by revans2 — COMMENTED** [review]
In regular Spark what does an empty iterator do?  Does it send to data to python and get no data back?

File: integration_tests/src/main/python/udf_cudf_test.py (hunk 1 above)
Comment (firestarman) [reply integration_tests/src/main/python/udf_cudf_test.py:48]: Updated the PR description for this change, I just explained it in the commit message only before. 
This change is to save the cudf tests execution time, especially for databricks, and it is unrelated to the IPC error.

**Comment by firestarman** [comment]
> In regular Spark what does an empty iterator do? Does it send to data to python and get no data back?

Per the code [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala#L88), no data will be sent to the Python side and I observed the UDF will not run for empty partitions.
But you can see the arrow writer still executes the `start `and `end `, similar with our GPU arrow writer's `constructor `and `close`. But ours seems did not send the right metadata for some reason, i am not sure. So the python side reads an illegal value.

**Review by revans2 — APPROVED** [review]
We might want a follow on issue to address the root cause, but this is fine for now.

**Comment by firestarman** [comment]
> We might want a follow on issue to address the root cause, but this is fine for now.

yes, i have re-opened this issue https://github.com/NVIDIA/spark-rapids/issues/750

**Comment by firestarman** [comment]
Closed the issue  https://github.com/NVIDIA/spark-rapids/issues/750 since I found  the root cause, pls go to the issue for details.
